fvm/transactionInvocator.go

```diff
@@ -160,17 +160,18 @@ func (i *TransactionInvocator) Process(
 		txError = fmt.Errorf("transaction invocation failed: %w", err)
 	}
 
-	if txError == nil {
-		txError = NewTransactionStorageLimiter().CheckLimits(env, sth.State().UpdatedAddresses())
+	// try to deduct fees even if there is an error.
+	feesError := i.deductTransactionFees(env, proc)
+	if feesError != nil {
+		txError = feesError
 	}
```

Comment:
I think we need env.Commit() to be after the fee deduction so we can merge register touches as well.